From cf35c678e121fe8e85cccf97a96140e7fcb806fa Mon Sep 17 00:00:00 2001
From: Romain Vimont <rom1v@videolabs.io>
Date: Wed, 10 Feb 2021 12:00:44 +0100
Subject: [PATCH 01/13] medialib: disable if already in use

The medialib does not properly support multiple simultaneous clients, so
running two VLC with medialib enabled would cause issues.

If the medialibrary is already in use, behave as if --no-media-library
was passed.
---
 modules/misc/medialibrary/medialibrary.cpp | 31 ++++++++++++++++------
 modules/misc/medialibrary/medialibrary.h   |  5 +++-
 2 files changed, 27 insertions(+), 9 deletions(-)

diff --git a/modules/misc/medialibrary/medialibrary.cpp b/modules/misc/medialibrary/medialibrary.cpp
index fb3d158706..1c8badd151 100644
--- a/modules/misc/medialibrary/medialibrary.cpp
+++ b/modules/misc/medialibrary/medialibrary.cpp
@@ -375,11 +375,24 @@ void MediaLibrary::onRescanStarted()
     m_vlc_ml->cbs->pf_send_event( m_vlc_ml, &ev );
 }
 
-MediaLibrary::MediaLibrary( vlc_medialibrary_module_t* ml )
-    : m_vlc_ml( ml )
+MediaLibrary* MediaLibrary::create( vlc_medialibrary_module_t* vlc_ml )
 {
-    m_ml.reset( NewMediaLibrary() );
+    auto userDir = vlc::wrap_cptr( config_GetUserDir( VLC_USERDATA_DIR ) );
+    auto mlDir = std::string{ userDir.get() } + "/ml/";
+    auto dbPath = mlDir + "ml.db";
+    auto mlFolderPath = mlDir + "mlstorage/";
+    auto ml = NewMediaLibrary( dbPath.c_str(), mlFolderPath.c_str(), true );
+    if ( !ml )
+        return {};
+
+    return new MediaLibrary( vlc_ml, ml );
+}
 
+MediaLibrary::MediaLibrary( vlc_medialibrary_module_t* vlc_ml,
+                            medialibrary::IMediaLibrary* ml )
+    : m_vlc_ml( vlc_ml )
+    , m_ml( ml )
+{
     m_logger.reset( new Logger( VLC_OBJECT( m_vlc_ml ) ) );
     m_ml->setVerbosity( var_InheritBool( VLC_OBJECT( m_vlc_ml ), "ml-verbose" ) ?
                           medialibrary::LogLevel::Debug : medialibrary::LogLevel::Warning );
@@ -392,16 +405,13 @@ bool MediaLibrary::Init()
     if( m_initialized )
         return true;
 
-    auto userDir = vlc::wrap_cptr( config_GetUserDir( VLC_USERDATA_DIR ) );
-    std::string mlDir = std::string{ userDir.get() } + "/ml/";
-
     m_ml->registerDeviceLister( std::make_shared<vlc::medialibrary::DeviceLister>(
                                     VLC_OBJECT(m_vlc_ml) ), "smb://" );
     m_ml->addFileSystemFactory( std::make_shared<vlc::medialibrary::SDFileSystemFactory>(
                                     VLC_OBJECT( m_vlc_ml ), m_ml.get(), "file://") );
     m_ml->addFileSystemFactory( std::make_shared<vlc::medialibrary::SDFileSystemFactory>(
                                     VLC_OBJECT( m_vlc_ml ), m_ml.get(), "smb://") );
-    auto initStatus = m_ml->initialize( mlDir + "ml.db", mlDir + "/mlstorage/", this );
+    auto initStatus = m_ml->initialize( this );
     switch ( initStatus )
     {
         case medialibrary::InitializeResult::AlreadyInitialized:
@@ -1758,7 +1768,12 @@ static int Open( vlc_object_t* obj )
 
     try
     {
-        p_ml->p_sys = new MediaLibrary( p_ml );
+        auto ml = MediaLibrary::create( p_ml );
+        if ( !ml )
+            /* The medialib is already used by another process */
+            return VLC_EGENERIC;
+
+        p_ml->p_sys = ml;
     }
     catch ( const std::exception& ex )
     {
diff --git a/modules/misc/medialibrary/medialibrary.h b/modules/misc/medialibrary/medialibrary.h
index 8ba8d5c108..50f43a3604 100644
--- a/modules/misc/medialibrary/medialibrary.h
+++ b/modules/misc/medialibrary/medialibrary.h
@@ -134,7 +134,8 @@ private:
 class MediaLibrary : public medialibrary::IMediaLibraryCb
 {
 public:
-    MediaLibrary( vlc_medialibrary_module_t* ml );
+    static MediaLibrary* create( vlc_medialibrary_module_t* ml );
+
     bool Init();
     bool Start();
     int Control( int query, va_list args );
@@ -163,6 +164,8 @@ private:
     static medialibrary::SortingCriteria sortingCriteria( int sort );
 
 private:
+    MediaLibrary( vlc_medialibrary_module_t* vlc_ml, medialibrary::IMediaLibrary* ml );
+
     vlc_medialibrary_module_t* m_vlc_ml;
     std::unique_ptr<Logger> m_logger;
     std::unique_ptr<medialibrary::IMediaLibrary> m_ml;
-- 
2.27.0

